### PR TITLE
ready for review

### PR DESCRIPTION
**Pull request by skeydan**

### CODE REVIEW

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd [hunk 1]
```diff
@@ -0,0 +1,619 @@
...
+preview: images/bbb.png
+---
+
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+
+A bit more than a year ago, in his beautiful [guest post](https://blogs.rstudio.com/tensorflow/posts/2018-07-17-activity-detection/) Nick Strayer showed how to classify a set of everyday activities using smartphone-recorded gyroscope and accelerometer data. Accuracy was very good, but Nick went on to inspect classification results more closely. Were there activities more prone to misclassification than others? And how about those erroneous results: Did the network report them with equal, or less, confidence than those that were correct?
```
Comment (batpigandme) [inline _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]:
```suggestion
A bit more than a year ago, in his beautiful [guest post](https://blogs.rstudio.com/tensorflow/posts/2018-07-17-activity-detection/), Nick Strayer showed how to classify a set of everyday activities using smartphone-recorded gyroscope and accelerometer data. Accuracy was very good, but Nick went on to inspect classification results more closely. Were there activities more prone to misclassification than others? And how about those erroneous results: Did the network report them with equal, or less confidence than those that were correct?
```

Comment (batpigandme) [reply _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]: I added a comma to offset the "in...guest post", and removed the one between "less" and "confidence" in the last sentence. 
😠 Ugh, I hate when diffs don't work right!

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd [hunk 2]
```diff
@@ -0,0 +1,619 @@
...
+
+Technically, what we refer to when talking of _confidence_ in that manner is the _score_ obtained for the "winning" class after softmax activation ^[see [Winner takes all: A look at activations and cost functions](https://blogs.rstudio.com/tensorflow/posts/2018-10-11-activations-intro/)]. If that winning score is 0.9, we might say "the network is sure that's a gentoo penguin"; if it's 0.2, we'd instead conclude "to the network, neither option seemed fitting, but cheetah looked best".
```
Comment (batpigandme) [inline _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]:
```suggestion
Technically, what we speak of _confidence_ in that manner, we're referring to the _score_ obtained for the "winning" class after softmax activation ^[see [Winner takes all: A look at activations and cost functions](https://blogs.rstudio.com/tensorflow/posts/2018-10-11-activations-intro/)]. If that winning score is 0.9, we might say "the network is sure that's a gentoo penguin"; if it's 0.2, we'd instead conclude "to the network, neither option seemed fitting, but cheetah looked best".
```

Comment (batpigandme) [reply _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]: I played around a little with the wording in the first sentence above — feel free to take it or leave it as you like! (This is also one of those weird cases where "speak" sounds more natural to me than "talk," but I don't have a rationale).

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd [hunk 3]
```diff
@@ -0,0 +1,619 @@
...
+
+This use of "confidence" is convincing, but it has nothing to do with confidence -- or credibility, or prediction, what have you -- intervals. What we'd really like to be able to do is put distributions over the network's weights and make it _Bayesian_. Using _tfprobability_'s variational Keras-compatible layers, this is something we actually can do.
```
Comment (batpigandme) [inline _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]: > confidence -- or credibility, or prediction, what have you -- intervals.

The way this is now makes it sound like they're all different names for the same thing, is that what you're going for? (I genuinely don't know if that's the case in this context).

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+
+[Adding uncertainty estimates to Keras models with tfprobability](https://blogs.rstudio.com/tensorflow/posts/2019-06-05-uncertainty-estimates-tfprobability/) showed how to use a variational dense layer to obtain estimates of epistemic uncertainty. In this post, we modify the convnet used in Nick's post to be variational throughout. Before we start, let's quickly summarize the task.
```
Comment (batpigandme) [inline _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]:
```suggestion
[Adding uncertainty estimates to Keras models with tfprobability](https://blogs.rstudio.com/tensorflow/posts/2019-06-05-uncertainty-estimates-tfprobability/) shows how to use a variational dense layer to obtain estimates of epistemic uncertainty. In this post, we modify the convnet used in Nick's post to be variational throughout. Before we start, let's quickly summarize the task.
```

Comment (batpigandme) [reply _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]: Since the post is still there, it's still showing! 🙂

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd [hunk 5]
```diff
@@ -0,0 +1,619 @@
...
+
+## The task
+
+To create the [Smartphone-Based Recognition of Human Activities and Postural Transitions Data Set](http://archive.ics.uci.edu/ml/datasets/Smartphone-Based+Recognition+of+Human+Activities+and+Postural+Transitions) [@Reyes-Ortiz], the researchers had subjects walk, sit, stand, and transition from one of those activities to another. Meanwhile, two types of smartphone sensors were used to record motion data: _Accelerometers_ measure linear acceleration in three ^[or two, depending on the application] dimensions, while _gyroscopes_ are used to track angular velocity around the coordinate axes. Here are, from Nick's original post, the respective raw sensor data for six types of activities:
```
Comment (batpigandme) [inline _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]:
```suggestion
To create the [Smartphone-Based Recognition of Human Activities and Postural Transitions Data Set](http://archive.ics.uci.edu/ml/datasets/Smartphone-Based+Recognition+of+Human+Activities+and+Postural+Transitions) [@Reyes-Ortiz], the researchers had subjects walk, sit, stand, and transition from one of those activities to another. Meanwhile, two types of smartphone sensors were used to record motion data: _Accelerometers_ measure linear acceleration in three ^[or two, depending on the application] dimensions, while _gyroscopes_ are used to track angular velocity around the coordinate axes. Here are the respective raw sensor data for six types of activities from Nick's original post:
```

Comment (batpigandme) [reply _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]: Switched the last sentence to have "from Nick's original post" at the end

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd [hunk 6]
```diff
@@ -0,0 +1,619 @@
...
+$ experiment    <int> 11, 12, 15, 16, 32, 33, 42, 43, 52, 53, 56, 57, 11, …
+$ userId        <int> 6, 6, 8, 8, 16, 16, 21, 21, 26, 26, 28, 28, 6, 6, 8,…
+$ activity      <int> 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 7, 8, 8, 8, 8, 8, 8…
+$ data          <list> [<data.frame[185 x 6]>, <data.frame[151 x 6]>, <dat…
+$ activityName  <fct> STAND_TO_SIT, STAND_TO_SIT, STAND_TO_SIT, STAND_TO_S…
+$ observationId <int> 11, 12, 15, 16, 31, 32, 41, 42, 51, 52, 55, 56, 71, …
+```
+
+
+The code required to arrive at this stage (copied from Nick's post) may be found in the appendix at the bottom of this page.
```
Comment (batpigandme) [inline _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]:
```suggestion
The code required to arrive at this stage may be found in the appendix at the bottom of this page.
```

Comment (batpigandme) [reply _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]: I think that it's clear you're borrowing from Nick's post at this point!

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd [hunk 7]
```diff
@@ -0,0 +1,619 @@
...
+
+## Training pipeline
+
+The dataset in question is small enough to fit in memory -- but yours might not be, so it can't hurt to see some streaming in action. Besides, it's probably safe to say that with TensorFlow 2.0, [tfdatasets](https://tensorflow.rstudio.com/tools/tfdatasets/articles/introduction.html) pipelines are _the_ way to feed data to a model.
+
+Once the code listed in the appendix has run, the sensor data is to be found in `trainData$data`, a list column containing `data.frame`s where each row corresponds to a point in time and each column holds one of the measurements. However, not all time series (recordings) are of the same length; we thus follow the original post to pad all series to length `pad_size` (= 338). The expected shape of training batches will then be `(batch_size, pad_size, 6)`.
```
Comment (batpigandme) [inline _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]:
```suggestion
Once the code listed in the appendix has run, the sensor data can be found in `trainData$data`, a list column containing `data.frame`s where each row corresponds to a point in time and each column holds one of the measurements. However, not all time series (recordings) are of the same length; we thus follow the original post to pad all series to length `pad_size` (= 338). The expected shape of training batches will then be `(batch_size, pad_size, 6)`.
```

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd [hunk 8]
```diff
@@ -0,0 +1,619 @@
...
+  one_hot_classes() %>% 
+  tensor_slices_dataset()
+
+n_test <- nrow(test_data)
+test_dataset <- zip_datasets(test_x, test_y) %>%
+  dataset_batch(n_test)
+
+```
+
+Using `tfdatasets` does not mean we cannot run a quick sanity check on our data:
```
Comment (batpigandme) [inline _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]: The double-negative above sounds a little off. Perhaps something ~ "We'll run a quick sanity check on our data (even though it's from `tfdatasets`)." or something of the like?

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd [hunk 9]
```diff
@@ -0,0 +1,619 @@
...
+ [ 1.04166663  0.26944447  0.12777779 -0.26755899 -0.02779437 -0.1441642 ]
+ [ 1.0250001   0.27083334  0.15277778 -0.19639318  0.35094208 -0.16249016]],
+ shape=(338, 6), dtype=float64)
+```
+
+Now let's build the network.
+
+## A variational convnet
+
+We build on the straightforward convolutional architecture from Nick's post, just making minor modifications to kernel sizes and numbers of filters. We also throw out all dropout layers; no additional regularization is needed on top of the priors applied to the weights.
```
Comment (batpigandme) [inline _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]:
```suggestion
We build on the straightforward convolutional architecture from Nick's post, making minor modifications to kernel sizes and numbers of filters. We also throw out all dropout layers; no additional regularization is needed on top of the priors applied to the weights.
```

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd [hunk 10]
```diff
@@ -0,0 +1,619 @@
...
+
+# the NLL part
+nll_part <- function(y_true, y_pred) {
+    cat_dist <- tfd_one_hot_categorical(logits = y_pred)
+    nll <- - (cat_dist %>% tfd_log_prob(y_true) %>% tf$reduce_mean())
+    nll
+}
+```
+
+We train somewhat longer than Nick did in the original post, allowing for early stopping though.
```
Comment (batpigandme) [inline _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]:
```suggestion
We train somewhat longer than Nick did in the original post, allowing for early stopping. 
```

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd [hunk 11]
```diff
@@ -0,0 +1,619 @@
...
+  train_dataset,
+  epochs = 1000,
+  validation_data = test_dataset,
+  callbacks = list(
+    callback_early_stopping(patience = 10)
+  )
+)
+```
+
+While the overall loss declines linearly (and probably would for many more epochs), this is not the case for classification accuracy and the NLL part of the loss:
```
Comment (batpigandme) [inline _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]:
```suggestion
While the overall loss declines linearly (and probably would for many more epochs), this is not the case for classification accuracy or the NLL part of the loss:
```

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd [hunk 12]
```diff
@@ -0,0 +1,619 @@
...
+knitr::include_graphics("images/history.png")
+```
+
+Final accuracy is not as high as in the non-variational setup, though still not bad for a six-class problem. We see that without any additional regularization, there is very little overfitting to the training data.
+
+Now how do we obtain predictions from this model?
+
+## Probabilistic predictions
+
+Though we won't go into this here, it's good to know that we can't just access the output distributions; through their `kernel_posterior` attribute, we can access the hidden layers' posterior weight distributions as well.
```
Comment (batpigandme) [inline _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]:
```suggestion
Though we won't go into this here, it's good to know that we access more than just the output distributions; through their `kernel_posterior` attribute, we can access the hidden layers' posterior weight distributions as well. 
```

Comment (batpigandme) [reply _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]: ^^ Assuming that's what you mean (my original reading of it was you saying _We can't access the output distributions_, until I read the other part of the sentence)

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd [hunk 13]
```diff
@@ -0,0 +1,619 @@
...
+
+Final accuracy is not as high as in the non-variational setup, though still not bad for a six-class problem. We see that without any additional regularization, there is very little overfitting to the training data.
+
+Now how do we obtain predictions from this model?
+
+## Probabilistic predictions
+
+Though we won't go into this here, it's good to know that we can't just access the output distributions; through their `kernel_posterior` attribute, we can access the hidden layers' posterior weight distributions as well. 
+
+Given the small size of the test set, we compute all predictions at once. The predictions now are categorical distributions, one for each sample in the batch:
```
Comment (batpigandme) [inline _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]:
```suggestion
Given the small size of the test set, we compute all predictions at once. The predictions are now categorical distributions, one for each sample in the batch: 
```

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd [hunk 14]
```diff
@@ -0,0 +1,619 @@
...
+one_shot_preds
+```
+
+```
+tfp.distributions.OneHotCategorical(
+ "sequential_one_hot_categorical_OneHotCategorical_OneHotCategorical",
+ batch_shape=[69], event_shape=[6], dtype=float32)
+```
+
+We prefixed those predictions by `one_shot` to indicate their noisy nature: These are predictions obtained on a single pass through the network, all layer weights being sampled from their respective posteriors.
```
Comment (batpigandme) [inline _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]:
```suggestion
We prefixed those predictions with `one_shot` to indicate their noisy nature: These are predictions obtained on a single pass through the network, all layer weights being sampled from their respective posteriors.
```

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd [hunk 15]
```diff
@@ -0,0 +1,619 @@
...
+# A tibble: 2 x 5
+  correct count avg_mean avg_sd avg_mc_sd
+  <lgl>   <int>    <dbl>  <dbl>     <dbl>
+1 FALSE      19    0.775  0.380     0.237
+2 TRUE       50    0.879  0.264     0.183
+```
+
+They are; though perhaps not to the extent we might desire.
+
+With just six classes, we can as well inspect standard deviations on the individual prediction-target pairings level.
```
Comment (batpigandme) [inline _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]:
```suggestion
With just six classes, we can also inspect standard deviations on the individual prediction-target pairings level.
```

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd (hunk 2 above)
Comment (skeydan) [reply _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]: hey, thx! I think you mean "when we speak of", so I'll skip the autocommit and edit this manually

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd (hunk 3 above)
Comment (skeydan) [reply _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]: hm, no... of course i don't want to give the impression like to me these are the same...
I'll think about it... on first thought i felt like removing, but "credibility" makes for a good link to "Bayesian"...

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd (hunk 4 above)
Comment (skeydan) [reply _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]: lol!

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd (hunk 5 above)
Comment (skeydan) [reply _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]: haha, the usual :-)

File: _posts/2019-11-13-variational-convnet/variational_convnet.Rmd (hunk 12 above)
Comment (skeydan) [reply _posts/2019-11-13-variational-convnet/variational_convnet.Rmd]: oh yeah... you always spot the weird-ish phrases ...

**Comment by skeydan** [comment]
Thanks a lot Mara, that was very helpful!!
